src/charm.py

```diff
@@ -292,14 +295,16 @@ def _on_config_changed(self, _: ops.HookEvent, charm_state: CharmState) -> None:
         self.reconcile(charm_state)
         self._set_workload_version()
 
-    @inject_charm_state
-    def _on_relation_departed(self, event: RelationDepartedEvent, charm_state: CharmState) -> None:
+    @validate_charm_state
+    def _on_relation_departed(self, event: RelationDepartedEvent) -> None:
         """Handle Synapse peer relation departed event.
 
         Args:
             event: relation departed event.
-            charm_state: The charm state.
         """
+        charm_state = self.build_charm_state()
+        MASConfiguration.validate(self)
```

Comment:
Couldnt this be added to the self.build_charm_state()?

Reply:
I was thinking of spliting the charm state into multiple components so that they can be initialized separately, and ideally in the future the hooks can choose what components they need and only initialize those components instead of the whole charm state. So I wouldn't want to put them together into a wrapper method.

And also in this PR I didn't use the initialized MAS charm state so I use the `validate` method here, but in the next PR I changed this to `mas_configuration = MASConfiguration.from_charm(self)`